sdk/tests/conformance/textures/misc/texture-size-limit.html

```diff
@@ -125,43 +125,63 @@
-        + (t.maxSize + 1) + "x" + (t.maxSize + 1));
-    // width and height out of bounds for specified level
-    gl.texImage2D(target, (numLevels - 1), test.format, 4, 4, 0, test.format, test.type, null);
-    wtu.glErrorShouldBe(gl, gl.INVALID_VALUE, "width or height out of bounds for specified level: should generate INVALID_VALUE: level is "
-        + (numLevels - 1) + ", size is 4x4.");
 
+    // Probe to discover the max non-OOM level.
+    // For instance, on some drivers (at least Intel+Mesa) we can create
+    // a maxLevel L8 texture, but only a maxLevel-1 RGB8 texture.
+    var maxLevelsForFormat = numLevels;
+    while (true) {
+      gl.texImage2D(target, maxLevelsForFormat-1, test.format, 1, 1, 0, test.format, test.type, pixels);
+      var err = gl.getError();
+      if (err == gl.OUT_OF_MEMORY) {
+        debug("Probe failed for level=" + (maxLevelsForFormat-1) + ", reducing...");
+        maxLevelsForFormat -= 1;
+        if (!maxLevelsForFormat) {
+          testFailed("Failed to allocate any levels for format " + test.format);
+          return;
+        }
+        continue;
+      }
+      if (err) {
+        testFailed("Should not hit non-OOM errors during max level probing.");
+        return;
+      }
+      break;
+    }
+    var numTestLevels = Math.min(maxLevelsForFormat, t.maxLevel);
     for (var l = 0; l < numTestLevels; ++l) {
       // Do bottom levels first;
       var size = 1 << l;
-      var level = numLevels - l - 1;
+      var level = maxLevelsForFormat - l - 1;
       var otherDimension = t.target == gl.TEXTURE_2D ? 1 : size;
       var badSize = size * 2;
       var badOtherDimension = t.target == gl.TEXTURE_2D ? 1 : badSize;
       var pixels = new test.dataType(badSize * badOtherDimension * test.size);
       gl.texImage2D(target, level, test.format, size, otherDimension, 0, test.format, test.type, pixels);
       wtu.glErrorShouldBe(gl, gl.NO_ERROR, "there should be no error for level: " + level + " " + size + "x" + otherDimension);
-      gl.texImage2D(target, level, test.format, otherDimension, size, 0, test.format, test.type, pixels);
-      wtu.glErrorShouldBe(gl, gl.NO_ERROR, "there should be no error for level: " + level + " " + otherDimension + "x" + size);
-      gl.texImage2D(target, level, test.format, badSize, badOtherDimension, 0, test.format, test.type, pixels);
-      wtu.glErrorShouldBe(gl, gl.INVALID_VALUE, "should generate INVALID_VALUE for level: " + level + " " + badSize + "x" + badOtherDimension);
-      gl.texImage2D(target, level, test.format, badOtherDimension, badSize, 0, test.format, test.type, pixels);
-      wtu.glErrorShouldBe(gl, gl.INVALID_VALUE, "should generate INVALID_VALUE for level: " + level + " " + badOtherDimension + "x" + badSize);
+      if (badOtherDimension != badSize) {
+        gl.texImage2D(target, level, test.format, otherDimension, size, 0, test.format, test.type, pixels);
```

Comment:
Did you mean to use badOtherDimension here?

Reply:
Well, maybe not. I was misled by the `badOtherDimension != badSize` condition. Maybe if it's correct, the condition could be refactored? (maybe `t.target == gl.TEXTURE_2D`, because that seems to be equivalent)

Reply:
If you're going to remove the two tests that were below this one, I actually don't think badSize and badOtherDimension need to exist at all; it seems like `pixels` can be changed to `size * otherDimension * test.size` and the `bad` variables can be removed.

Reply:
I see you resolved this. :)